adding airgap scenario for mgmt-cluster (#307)
suse-edge/suse-edge.github.io: asciidoc/product/atip-management-cluster.adoc

There are three main steps to set up the management cluster using a declarative approach:

> . **xref:mgmt-cluster-image-preparation-connected[Image preparation for connected environments]**: The first step is to prepare the manifests and files with all the necessary configurations to be used in connected environments.

Reviewer: @alknopfler Have you used this reference elsewhere in the document You might have to change that accordingly.

Author: yes I did a replace and all references has been changed
![image](https://github.com/suse-edge/suse-edge.github.io/assets/9356454/3154df51-a3e0-4674-a9e9-c2a8dd88bad4)

        ** Config: Contains the configuration files to be used by the management cluster.
  - xref:mgmt-cluster-network-folder[Network folder]: The `network` folder contains the network configuration files to be used by the management cluster nodes.

. **xref:mgmt-cluster-image-preparation-airgap[Image preparation for air-gap environments]**: The step is to show the differences to prepare the manifests and files to be used in an air-gap scenario.
  - xref:mgmt-cluster-directory-structure-airgap[Directory structure for air-gap environments]: The directory structure must be modified to include the resources needed to run the management cluster in an air-gap environment.
  - xref:mgmt-cluster-image-definition-file-airgap[Modifications in the definition file]: The `mgmt-cluster.yaml` file must be modified to include the `embeddedArtifactRegistry` section with the `images` field set to all container images to be included into the EIB output image.
  - xref:mgmt-cluster-custom-folder-airgap[Modifications in the custom folder]: The `custom` folder must be modified to include the resources needed to run the management cluster in an air-gap environment.

>         ** Register script: The `custom/scripts/99-register.sh` script must be removed when use an air-gap environment.

Reviewer: ```suggestion
        ** Register script: The `custom/scripts/99-register.sh` script must be removed when you use an air-gap environment.
```

- The `mgmt-cluster.yaml` file must be modified to include the `embeddedArtifactRegistry` section with the `images` field set to all container images to be included into the EIB output image.
- The `custom/scripts/99-register.sh` script must be removed when use an air-gap environment.
- The `custom/files/airgap-resources.tar.gz` file must be included in the `custom/files` folder with all the resources needed to run the management cluster in an air-gap environment.
- The `custom/scripts/99-mgmt-setup.sh` script must be modified to extract and copy the `airgap-resources.tar.gz` file to the final location.
- The `custom/files/metal3.sh` script must be modified to use the local resources included in the `airgap-resources.tar.gz` file instead of downloading them from the internet.

[#mgmt-cluster-directory-structure-airgap]
==== Directory structure for air-gap environments

> The directory structure for air-gap environments is the same as for connected environments, with the following differences will be explained below:

Reviewer: ```suggestion
The directory structure for air-gap environments is the same as for connected environments, with the differences explained as follows:
```

[,console]
----
eib
|-- base-images
|   |-- SLE-Micro.x86_64-5.5.0-Default-SelfInstall-GM2.install.iso
|-- custom
|   |-- files
|   |   |-- airgap-resources.tar.gz
|   |   |-- basic-setup.sh
|   |   |-- metal3.sh
|   |   |-- mgmt-stack-setup.service
|   |   |-- rancher.sh
|   |-- scripts
|       |-- 99-alias.sh
|       |-- 99-mgmt-setup.sh
|-- kubernetes
|   |-- config
|   |   |-- server.yaml
|   |-- helm
|   |   |-- values
|   |       |-- certmanager.yaml
|   |       |-- metal3.yaml
|   |       |-- neuvector.yaml
|   |       |-- rancher.yaml
|   |-- manifests
|       |-- neuvector-namespace.yaml
|-- mgmt-cluster.yaml
|-- network
    |-- mgmt-cluster-network.yaml
----

[NOTE]
====

> The image `SLE-Micro.x86_64-5.5.0-Default-SelfInstall-GM2.install.iso` must be downloaded from the https://scc.suse.com/[SUSE Customer Center] or the https://www.suse.com/download/sle-micro/[SUSE Download page], and it must be located under the `base-images` folder.

Reviewer: A question: If it is air-gap,without internet, how can the image be downloaded from the pages?

Author: It will be previously from their laptop. I will add a note to explain that

You should check the SHA256 checksum of the image to ensure it has not been tampered with. The checksum can be found in the same location where the image was downloaded.

An example of the directory structure can be found in the https://github.com/suse-edge/atip[SUSE Edge GitHub repository under the "telco-examples" folder].
====

[#mgmt-cluster-image-definition-file-airgap]
==== Modifications in the definition file

The `mgmt-cluster.yaml` file must be modified to include the `embeddedArtifactRegistry` section with the `images` field set to all container images to be included into the EIB output image. The `images` field must contain the list of all container images to be included in the output image. The following is an example of the `mgmt-cluster.yaml` file with the `embeddedArtifactRegistry` section included:

[,yaml]
----
apiVersion: 1.0
image:
  imageType: iso
  arch: x86_64
  baseImage: SLE-Micro.x86_64-5.5.0-Default-SelfInstall-GM2.install.iso
  outputImageName: eib-mgmt-cluster-image.iso
operatingSystem:
  isoConfiguration:
    installDevice: /dev/sda
  users:
  - username: root

>     encryptedPassword: ${ROOT_PASSWORD}

Reviewer 2: Not in this PR but in future we may want to adjust the examples to use a non-root user with sudo instead - EIB doesn't actually provide any API for this AFAICS, so I ended up using a [script instead](https://github.com/suse-edge/metal3-demo/blob/main/roles/edge-image-builder/tasks/main.yml#L159)

	kind: Feature
	metadata:
	  name: embedded-cluster-api
	spec:
	  value: false
	EOF

  # Disable Rancher webhooks for CAPI
  ${KUBECTL} delete mutatingwebhookconfiguration.admissionregistration.k8s.io mutating-webhook-configuration
  ${KUBECTL} delete validatingwebhookconfigurations.admissionregistration.k8s.io validating-webhook-configuration
  ${KUBECTL} wait --for=delete namespace/cattle-provisioning-capi-system --timeout=300s
fi

# Deploy CAPI
if [ $(${KUBECTL} get pods -n ${METAL3_CAPISYSTEMNAMESPACE} -o name | wc -l) -lt 1 ]; then

  # Try this command 3 times just in case, stolen from https://stackoverflow.com/a/33354419
  if ! (r=3; while ! /opt/mgmt/bin/clusterctl init \
    --core "cluster-api:v${METAL3_CAPICOREVERSION}"\
    --infrastructure "metal3:v${METAL3_CAPIMETAL3VERSION}"\
    --bootstrap "${METAL3_CAPIPROVIDER}:v${METAL3_CAPIRKE2VERSION}"\
    --control-plane "${METAL3_CAPIPROVIDER}:v${METAL3_CAPIRKE2VERSION}"\
    --config /root/cluster-api/clusterctl.yaml ; do
            ((--r))||exit
            echo "Something went wrong, let's wait 10 seconds and retry"
            sleep 10;done) ; then
      echo "clusterctl failed"
      exit 1

>   fi

Reviewer 2: Is there a reason we need this retry for clusterctl?

If we're going to do it I think it would be cleaner to do something like:

```
retries=3
for _ in $(seq "$retries"); do
...
done
```

Author: I'd like to keep as is because it comes from SV. I will talk with Edu just to improve this part in all places

fi

# Clean up the lock cm

${KUBECTL} delete configmap ${METAL3LOCKCMNAME} -n ${METAL3LOCKNAMESPACE}
----

- The `custom/files/airgap-resources.tar.gz` file must be included in the `custom/files` folder with all the resources needed to run the management cluster in an air-gap environment. This file must be prepared manually downloading all resources and compressing them into this single file. The `airgap-resources.tar.gz` file contains the following resources:
+
[,console]
----
|-- clusterctl
|-- clusterctl.yaml
|-- overrides
    |-- bootstrap-rke2
    |   |-- v0.2.6
    |       |-- bootstrap-components.yaml
    |       |-- metadata.yaml
    |-- cluster-api
    |   |-- v1.6.2
    |       |-- core-components.yaml
    |       |-- metadata.yaml
    |-- control-plane-rke2
    |   |-- v0.2.6
    |       |-- control-plane-components.yaml
    |       |-- metadata.yaml
    |-- infrastructure-metal3
        |-- v1.6.0
            |-- cluster-template.yaml
            |-- infrastructure-components.yaml
            |-- metadata.yaml
----

> The `clusterctl.yaml` file contains the configuration to specify the images location and the overrides to be used by the `clusterctl` tool. The `overrides` folder contains yaml files manifests to be used instead of downloading them from the internet.

Reviewer: ```suggestion
The `overrides` folder contains `yaml` file manifests to be used instead of downloading them from the internet.
```
